src/d3a/models/myco_matcher/external_matcher.py

```diff
@@ -1,140 +1,179 @@
+            if not isinstance(ex, MycoValidationException):
+                logging.exception("Bid offer pair matching failed.")
+
+        self.myco_ext_conn.publish_json(channel, response_data)
 
-    def get_simulation_id(self, message):
-        """Publish the simulation id to the Myco client."""
+    def publish_simulation_id(self, message):
+        """Publish the simulation id to the redis myco client.
+
+        At the moment the id of the simulations run by the cli is set as ""
+        however, this function guarantees that the myco is aware of the running collaboration id
+        regardless of the value set in d3a.
+        """
 
         channel = "external-myco/get-simulation-id/response"
         self.myco_ext_conn.publish_json(channel, {"simulation_id": self.simulation_id})
 
     def publish_event_tick_myco(self):
         """Publish the tick event to the Myco client."""
 
-        channel = f"external-myco/{d3a.constants.COLLABORATION_ID}/response/events/"
         data = {"event": "tick"}
-        self.myco_ext_conn.publish_json(channel, data)
+        self.myco_ext_conn.publish_json(self.events_channel, data)
 
-    def publish_market_cycle_myco(self):
+    def publish_event_market_myco(self):
         """Publish the market event to the Myco client."""
 
-        channel = f"external-myco/{d3a.constants.COLLABORATION_ID}/response/events/"
         data = {"event": "market"}
-        self.myco_ext_conn.publish_json(channel, data)
+        self.myco_ext_conn.publish_json(self.events_channel, data)
 
     def publish_event_finish_myco(self):
         """Publish the finish event to the Myco client."""
 
-        channel = f"external-myco/{d3a.constants.COLLABORATION_ID}/response/events/"
         data = {"event": "finish"}
-        self.myco_ext_conn.publish_json(channel, data)
+        self.myco_ext_conn.publish_json(self.events_channel, data)
 
     def calculate_match_recommendation(self, bids, offers, current_time=None):
         pass
+
+    def update_area_uuid_markets_mapping(self, area_uuid_markets_mapping: Dict) -> None:
+        """Interface for updating the area_uuid_markets_mapping mapping."""
+        self.area_uuid_markets_mapping.update(area_uuid_markets_mapping)
+
+    @staticmethod
+    def _get_bids_offers(market: Market, filters: Dict):
+        """Get bids and offers from market, apply filters and return serializable lists."""
+
+        bids, offers = market.open_bids_and_offers
+        bids_list = list(bid.serializable_dict() for bid in bids.values())
+        filtered_offers_energy_type = filters.get("energy_type", None)
+        if filtered_offers_energy_type:
+            offers_list = list(
+                offer.serializable_dict() for offer in offers.values()
+                if offer.attributes and
+                offer.attributes.get("energy_type") == filtered_offers_energy_type)
+        else:
+            offers_list = list(
+                offer.serializable_dict() for offer in offers.values())
```

Comment:
Sorry if I'm getting a bit confused at this point.
1. The myco-api-client requests open bids and offers from d3a.
2. d3a answers with a list of open bids/offers

Now we want to add some filtering. However, shouldn't this filtering be done in the matching algorithm? In theory, `d3a` shouldn't know anything about filtering: it should just validate that the matches between bids and offer are valid. Right?

Reply:
![image-20210423-190639](https://user-images.githubusercontent.com/38125715/124959638-1757b300-e024-11eb-9604-528c02a63086.png)
The requirement states that myco can request and filter orders and then it would pass them to the matching algorithm, I can see some business goals behind this
anw this also favors the communication part, we only send the minimum size of needed data